src/Silverback.Integration/Messaging/Diagnostics/ActivityExtensions.cs

```diff
@@ -49,33 +50,50 @@ public static void SetMessageHeaders(this Activity activity, MessageHeaderCollec
             }
         }
 
-        // See https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
-        public static void InitFromMessageHeaders(this Activity activity, MessageHeaderCollection headers)
+        public static void SetTraceIdAndState(this Activity activity, string? traceId, string? traceState)
         {
-            string? traceId = headers.GetValue(DefaultMessageHeaders.TraceId);
-
             if (!string.IsNullOrEmpty(traceId))
             {
-                // This will reflect, that the current activity is a child of the activity
-                // which is represented in the message.
                 activity.SetParentId(traceId);
 
-                string? traceState = headers.GetValue(DefaultMessageHeaders.TraceState);
                 if (!string.IsNullOrEmpty(traceState))
                 {
                     activity.TraceStateString = traceState;
                 }
+            }
+        }
 
-                // The baggage is parsed last, so if it fails to be deserialized the rest is still setup.
-                // We expect baggage to be empty by default.
-                // Only very advanced users will be using it in near future, we encourage them to keep baggage small (few items).
-                string? baggage = headers.GetValue(DefaultMessageHeaders.TraceBaggage);
-                if (baggage != null)
-                {
-                    var baggageItems = ActivityBaggageSerializer.Deserialize(baggage);
-                    AddBaggageRange(activity, baggageItems);
-                }
+        public static void AddMessageId(this Activity activity, IBrokerMessageIdentifier? messageId)
```

Comment:
The `IBrokerMessageIdentifier` represents the Kafka offset. In Silverback we actually got 3 different identifiers when dealing with Kafka:
* the record offset (`IBrokerMessageIdentifier` / `IKafkaOffset`)
* the record key (see `x-kafka-message-key` header added by Silverback)
* the `x-message-id` header added by Silverback to all produced messages (if the header wasn't submitted, the consumer will populate it with the kafka key)

Calling the `IBrokerMessageIdentifier` argument `messageId` is confusing.

But what's the value we actually want to store in the activity? Is that the offset? Beware that no such construct exist in MQTT (or Rabbit). For the logging part I differentiate the information that gets logged according to the broker type. See https://github.com/BEagle1984/silverback/blob/master/src/Silverback.Integration.Kafka/Messaging/Configuration/KafkaBrokerOptionsConfigurator.cs etc.

Reply:
Not sure I understand correctly. Is `IBrokerMessageIdentifier` specific to Kafka? I thought that the Broker Implementation will store in it whatever is needed to identify a message? I don't really care what the content is, as long as it identifies the message (I just found the @ notation convenient, but we can use something else).

I would like to have a value that identifies the message exactly. I agree that this probably has to be implemented per Broker (and I assumed this is what the `IBrokerMessageIdentifier` is for).

Should we add some kind of `IBrokerActivityEnricher`? If so, we could possibly get rid of the configuration stuff you linked above, because once it is on the activity, it can automatically be logged with every log entry.

Reply:
`IBrokerMessageIdentifier` is generic but with MQTT  there isn't such thing like the offset and thus there's no meaningful value that can be stored in there. In order for Silverback to work I generate a new `Guid` in the consumer but adding that value to the trace is of no use (there is no correlation between producer and consumer and not even between two different consumers processing the same message).

Are the tags gonna be standardized at company level?

I'd personally call the Kafka offset, well, "offset" instead of "message id" and add it only for Kafka. You could store also the message-id from the header (always provided at least for SB-to-SB communication) and in case the kafka key.
What do you think?

I understand it's a bit more work to making it extensible via a mechanism like `IBrokerActivityEnricher`, that should be bound to a certain endpoint type (much like the `InboundLogger`, see `https://github.com/BEagle1984/silverback/blob/master/src/Silverback.Integration/Diagnostics/LoggerCollection.cs`) but that would really be the correct implementation in my opinion, leaving open the possibility to decide which information we store per each broker type (having different tag names as well).

Reply:
Yeah, I agree that the `IBrokerActivityEnricher` would be the best solution by far. 
Regarding the tag names: I would follow the spec here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md as closely as possible.

Reply:
Ah, nice! There's a kind of standard then. 😄  We could also provide quite a few extra tags with minimal effort (broker type, etc.) but I don't know how useful they are.
Then yes, let's stick to the message-id. We just have to decide if we want the offset, the kafka key or the message id header. 😝